### ISSUE TITLE
The MacOS compiler is generating illegal instruction for the division of c10::complex

### ISSUE THREAD

**Issue by zasdfgbnm** [pytorch/pytorch#39123]
The MacOS compiler is generating illegal instruction for the division of `c10::complex` numbers. Search `AT_DISPATCH_COMPLEX_TYPES_MAC_STD` in the code to find and remove the workaround for `logspace_cpu_out` and `div_kernel`.

**Comment by anjali411** [comment]
reopening the issue as it looks unresolved

**Comment by ezyang** [comment]
You mean, it's making an AVX2 instruction when it shouldn't?

**Comment by zasdfgbnm** [comment]
@ezyang The error is SIGILL (illegal instruction), but I don't know which instructions it is generating. But I tried to use `#ifdef __APPLE__` to replace `cpu_kernel_vec` with `cpu_kernel`, it does not help, so seems like it is not a problem of `Vec256`.

**Comment by ezyang** [comment]
@zasdfgbnm Remember last time where you generated a weak symbol in the vectorized codegen and it clobbered a non-vectorized compilation of the symbol? It could be that case all over again. Need to make sure we're not defining any static data in the vec headers. If you can gdb the failing instruction that should also give you more info (use `disas` to find out more about the assembly in the end.)

**Comment by zasdfgbnm** [comment]
@ezyang Yeah, you are right. Thanks for the hint! I am not using static data, but the problem is, weak symbols for `operator /=` of `c10::complex` are generated. I added a `__attribute__((always_inline))` then the problem is solved.

I messed up the original PR (I created a PR from the branch of https://github.com/pytorch/pytorch/pull/39553 to the branch of #39300, and merged it, then #39300 becomes no longer mergeable to master.), please see the new PR https://github.com/pytorch/pytorch/pull/39553.

But there is one thing that really makes me worried: in general, any function could potentially be built against AVX2, and generate weak symbols, and then called and failed with a SIGILL, this really depends on how the compiler wants to handle it. So I feel it would be good to separate `libtorch_avx2.so` which hides symbols for internal use, just like `libtorch_cuda.so`, this might be safer.

**Comment by ezyang** [comment]
> But there is one thing that really makes me worried: in general, any function could potentially be built against AVX2, and generate weak symbols, and then called and failed with a SIGILL, this really depends on how the compiler wants to handle it. So I feel it would be good to separate libtorch_avx2.so which hides symbols for internal use, just like libtorch_cuda.so, this might be safer.

Yes, I think this may generally be a good idea. File an issue for it perhaps. Also cc @colesbury

**Comment by colesbury** [comment]
It's been a while, but I think I originally tried having the architecture specific kernels in a separate library. I ran into problems with the Facebook internal build. I think it was because buck didn't support the necessary circular dependencies between libraries (libtorch needs to depend on the AVX kernels and vice versa). You're welcome to try to work around it.

Regarding weak symbols, we should also try to link the AVX/AVX2 object files last so that symbols from the rest of libtorch are preferred.
